modules/layers/src/text-layer/utils.js

```diff
@@ -64,3 +68,75 @@ export function buildMapping({
     canvasHeight: nextPowOfTwo(yOffset + (row + 1) * rowHeight)
   };
 }
+
+export function updateRange({array, startIndex, endIndex, data}) {
+  for (let i = startIndex; i < Math.min(endIndex, array.length); i++) {
+    const object = array[i];
+    Object.assign(object, data);
+  }
+}
+
+export function transformText(text, iconMapping, transformLetter, transformedData) {
+  let startIndex = transformedData.length;
+
+  const letters = Array.from(text);
+
+  let offsetLeft = 0;
+  let offsetTop = 0;
+  let lineHeight = 0;
+
+  // width and height of the text
+  const size = [0, 0];
+
+  letters.forEach((letter, i) => {
+    if (letter === '\n') {
+      size[0] = Math.max(offsetLeft, size[0]);
+
+      updateRange({
```

Comment:
This retro update is very hard to read. Just break your lines like

```js
letters.split('\n').forEach(...)
```

And you can access the line length as you build the data.